packages/handlersjs-core/lib/handlers/waterfall-handler.ts

```diff
@@ -36,7 +36,7 @@ export class WaterfallHandler<T, S> extends Handler<T, S> {
     input: T,
     intermediateOutput: S,
     handlers: Handler<T, S>[],
-  ): Observable<Handler<T, S>> {
+  ): Observable<Handler<T, S> | undefined> {
```

Comment:
Instead of returning undefined, I would switchmap after the last map, throwing an error if the result is undefined.